### PR TITLE
Swagger Documentation of EntityFieldConfigurationApiResources.java and LoanChargesApiResource.java

### PR DESCRIPTION
**Pull request by luckyman20**
GCI Task : 5696322638184448

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/EntityFieldConfigurationApiResources.java [hunk 1]
```diff
@@ -83,7 +75,9 @@ public EntityFieldConfigurationApiResources(final PlatformSecurityContext contex
 	@GET
 	@Consumes({ MediaType.APPLICATION_JSON })
 	@Produces({ MediaType.APPLICATION_JSON })
-	public String getAddresses(@PathParam("entity") final String entityname, @Context final UriInfo uriInfo) {
+	@ApiOperation(value = "Lists Entity Field Configuration", httpMethod = "GET", notes = "It Lists Entity Field Configuration")
+    @ApiResponses({@ApiResponse(code = 200, message = "OK", response = EntityFieldConfigurationApiResourcesSwagger.GetEntityFieldConfigurationResponse.class, responseContainer = "List")})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/EntityFieldConfigurationApiResources.java]: 1. Indentation is not correct!
2. Refractor the name of Inner Class to GetFieldConfigurationEntityResponse

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/EntityFieldConfigurationApiResourcesSwagger.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ * under the License.
+ */
+package org.apache.fineract.portfolio.address.api;
+
+import io.swagger.annotations.ApiModelProperty;
+
+/**
+ * Created by Chirag Gupta on 12/01/17.
+ */
+public class EntityFieldConfigurationApiResourcesSwagger {
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/EntityFieldConfigurationApiResourcesSwagger.java]: make this class "final"

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 3]
```diff
@@ -96,7 +89,9 @@ private boolean is(final String commandParam, final String commandValue) {
     @GET
     @Consumes({ MediaType.APPLICATION_JSON })
     @Produces({ MediaType.APPLICATION_JSON })
-    public String retrieveAllLoanCharges(@PathParam("loanId") final Long loanId, @Context final UriInfo uriInfo) {
+    @ApiOperation(value = "List Loan Charges", httpMethod = "GET", notes = "It lists all the Loan Charges specific to a Loan \n\n" + "Example Requests:\n" + "\n" + "loans/1/charges\n" + "\n" + "\n" + "loans/1/charges?fields=name,amountOrPercentage")
+    @ApiResponses({@ApiResponse(code = 200, message = "OK", responseContainer = "List", response = LoanChargesApiResourceSwagger.GetLoanChargesResponse.class)})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to GetLoansLoanIdChargesResponse

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 4]
```diff
@@ -110,7 +105,9 @@ public String retrieveAllLoanCharges(@PathParam("loanId") final Long loanId, @Co
     @Path("template")
     @Consumes({ MediaType.APPLICATION_JSON })
     @Produces({ MediaType.APPLICATION_JSON })
-    public String retrieveTemplate(@PathParam("loanId") final Long loanId, @Context final UriInfo uriInfo) {
+    @ApiOperation(value = "Retrieve Loan Charges Template", httpMethod = "GET", notes = "This is a convenience resource. It can be useful when building maintenance user interface screens for client applications. The template data returned consists of any or all of:\n" + "\n" + "Field Defaults\n" + "Allowed Value Lists\n" + "Example Request:\n" + "\n" + "loans/1/charges/template\n" + "\n")
+    @ApiResponses({@ApiResponse(code = 200, message = "OK", response = LoanChargesApiResourceSwagger.GetLoanChargeTemplateResponse.class )})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to GetLoansLoanIdChargesTemplateResponse

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 5]
```diff
@@ -126,7 +123,9 @@ public String retrieveTemplate(@PathParam("loanId") final Long loanId, @Context
     @Path("{chargeId}")
     @Consumes({ MediaType.APPLICATION_JSON })
     @Produces({ MediaType.APPLICATION_JSON })
-    public String retrieveLoanCharge(@PathParam("loanId") final Long loanId, @PathParam("chargeId") final Long loanChargeId,
+    @ApiOperation(value = "Retrieve a Loan Charge", httpMethod = "GET", notes = "Retrieves Loan Charge according to the Loan ID and Charge ID" + "Example Requests:\n" + "\n" + "/loans/1/charges/1\n" + "\n" + "\n" + "/loans/1/charges/1?fields=name,amountOrPercentage")
+    @ApiResponses({@ApiResponse(code = 200, message = "OK", response = LoanChargesApiResourceSwagger.GetLoanChargesResponse.class)})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to GetLoansLoanIdChargesChargeIdResponse

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 6]
```diff
@@ -145,8 +144,11 @@ public String retrieveLoanCharge(@PathParam("loanId") final Long loanId, @PathPa
     @POST
     @Consumes({ MediaType.APPLICATION_JSON })
     @Produces({ MediaType.APPLICATION_JSON })
-    public String executeLoanCharge(@PathParam("loanId") final Long loanId, @QueryParam("command") final String commandParam,
-            final String apiRequestBodyAsJson) {
+    @ApiOperation(value = "Create a Loan Charge", httpMethod = "POST", notes = "It Creates a Loan Charge")
+    @ApiImplicitParams({@ApiImplicitParam(value = "body", required = true, paramType = "body", dataType = "body", format = "body", dataTypeClass = LoanChargesApiResourceSwagger.PostLoanChargesRequest.class)})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to PostLoansLoanIdChargesRequest

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 7]
```diff
@@ -145,8 +144,11 @@ public String retrieveLoanCharge(@PathParam("loanId") final Long loanId, @PathPa
...
+    @ApiResponses({@ApiResponse(code = 200, message = "OK", response = LoanChargesApiResourceSwagger.PostLoanChargesResponse.class)})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to PostLoansLoanIdChargesResponse

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 8]
```diff
@@ -166,8 +168,11 @@ public String executeLoanCharge(@PathParam("loanId") final Long loanId, @QueryPa
     @Path("{chargeId}")
     @Consumes({ MediaType.APPLICATION_JSON })
     @Produces({ MediaType.APPLICATION_JSON })
-    public String updateLoanCharge(@PathParam("loanId") final Long loanId, @PathParam("chargeId") final Long loanChargeId,
-            final String apiRequestBodyAsJson) {
+    @ApiOperation(value = "Update a Loan Charge", httpMethod = "PUT", notes = "Currently Loan Charges may be updated only if the Loan is not yet approved")
+    @ApiImplicitParams({@ApiImplicitParam(value = "body", required = true, paramType = "body", dataType = "body", format = "body", dataTypeClass = LoanChargesApiResourceSwagger.PutLoanChargesRequest.class)})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to PutLoansLoanIdChargesChargeIdRequest

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 9]
```diff
@@ -166,8 +168,11 @@ public String executeLoanCharge(@PathParam("loanId") final Long loanId, @QueryPa
...
+    @ApiResponses({@ApiResponse(code = 200, message = "OK", response = LoanChargesApiResourceSwagger.PutLoanChargesResponse.class)})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to PutLoansLoanIdChargesChargeIdResponse

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 10]
```diff
@@ -181,8 +186,11 @@ public String updateLoanCharge(@PathParam("loanId") final Long loanId, @PathPara
     @Path("{chargeId}")
     @Consumes({ MediaType.APPLICATION_JSON })
     @Produces({ MediaType.APPLICATION_JSON })
-    public String executeLoanCharge(@PathParam("loanId") final Long loanId, @PathParam("chargeId") final Long loanChargeId,
-            @QueryParam("command") final String commandParam, final String apiRequestBodyAsJson) {
+    @ApiOperation(value = "Pay Loan Charge", httpMethod = "POST", notes = "Loan Charge will be paid if the loan is linked with a savings account")
+    @ApiImplicitParams({@ApiImplicitParam(value = "body", required = true, paramType = "body", dataType = "body", format = "body", dataTypeClass = LoanChargesApiResourceSwagger.PostLoanChargeIdRequest.class)})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to PostLoansLoanIdChargesChargeIdRequest

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 11]
```diff
@@ -181,8 +186,11 @@ public String updateLoanCharge(@PathParam("loanId") final Long loanId, @PathPara
...
+    @ApiResponses({@ApiResponse(code = 200, message = "OK", response = LoanChargesApiResourceSwagger.PostLoanChargeIdResponse.class)})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to PostLoansLoanIdChargesChargeIdResponse

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java [hunk 12]
```diff
@@ -203,7 +211,9 @@ public String executeLoanCharge(@PathParam("loanId") final Long loanId, @PathPar
     @Path("{chargeId}")
     @Consumes({ MediaType.APPLICATION_JSON })
     @Produces({ MediaType.APPLICATION_JSON })
-    public String deleteLoanCharge(@PathParam("loanId") final Long loanId, @PathParam("chargeId") final Long loanChargeId) {
+    @ApiOperation(value = "Delete a Loan Charge", httpMethod = "DELETE", notes = "Note: Currently, A Loan Charge may only be removed from Loans that are not yet approved.")
+    @ApiResponses({@ApiResponse(code = 200, message = "OK", response = LoanChargesApiResourceSwagger.DeleteLoanChargeResponse.class)})
```
Comment (Sanyam96) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanChargesApiResource.java]: Refractor the inner class name to DeleteLoansLoanIdChargesChargeIdResponse

**Comment by Sanyam96** [comment]
Chirag, 
Follow proper naming conventions for all the inner classes!

**"MethodType+baseAPIRESPONSESLUG+methodSLUG+Request/Response"**

Example: 
**-->** for **Get** Method of EntityFieldConfigurationApiResources.java
**-->** the **BaseAPIResponseSlug** is: /fieldconfiguration/{entity} -> **FieldConfigurationEntity** [https://github.com/luckyman000786/incubator-fineract/blob/351aecaf17e0619c30e4fce2f4bf9917f3a57e25/fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/EntityFieldConfigurationApiResources.java#L43] (Line 43)
**-->** in this Get method there's no slug (Line 75) [https://github.com/luckyman000786/incubator-fineract/blob/351aecaf17e0619c30e4fce2f4bf9917f3a57e25/fineract-provider/src/main/java/org/apache/fineract/portfolio/address/api/EntityFieldConfigurationApiResources.java#L75]
**-->** we are expecting response so -> **Response**

so the inner class name should be **GetFieldConfigurationEntityResponse**

**Comment by luckyman20** [comment]
@Sanyam96  Updated the PR

**Pull request closed without merging**